webapp/tests/karma/unit/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+      });
+    });
+
+    it('sets localStorage values', () => {
+      storageGetItem.withArgs('medic-greg-telemetry-db').returns(undefined);
+      storageGetItem.withArgs('medic-greg-telemetry-date').returns(undefined);
+
+      pouchDb.post = sinon.stub().resolves();
+
+      return service.record('test', 1).then(() => {
+        chai.expect(storageSetItem.callCount).to.equal(2);
+        chai
+          .expect(storageSetItem.args[0][0])
+          .to.equal('medic-greg-telemetry-db');
+        chai
+          .expect(storageSetItem.args[0][1])
+          .to.match(/medic-user-greg-telemetry-/); // ends with a UUID
+        chai
+          .expect(storageSetItem.args[1][0])
+          .to.equal('medic-greg-telemetry-date');
+        chai.expect(storageSetItem.args[1][1]).to.equal(NOW);
+      });
+    });
+
+    it('aggregates once a month and resets the db', () => {
+      storageGetItem.withArgs('medic-greg-telemetry-db').returns('dbname');
+      storageGetItem.withArgs('medic-greg-telemetry-date').returns(
+        moment()
+          .subtract(5, 'weeks')
+          .valueOf()
+          .toString()
+      );
+
+      pouchDb.post = sinon.stub().resolves();
+      pouchDb.query = sinon.stub().resolves({
+        rows: [
+          {
+            key: 'foo',
+            value: 'stats',
+          },
+          {
+            key: 'bar',
+            value: 'more stats',
+          },
+        ],
+      });
+      DB.info.resolves({ some: 'stats' });
+      DB.put.resolves();
+      pouchDb.destroy = sinon.stub().resolves();
+
+      $window.navigator.userAgent = 'Agent Smith';
+      $window.navigator.hardwareConcurrency = 4;
+      $window.screen = {
+        availWidth: 768,
+        availHeight: 1024,
+      };
+
+      return service.record('test', 1).then(() => {
+        chai.expect(pouchDb.post.callCount).to.equal(1);
+        chai
+          .expect(pouchDb.post.args[0][0])
+          .to.deep.include({ key: 'test', value: 1 });
+        chai.expect(DB.put.callCount).to.equal(1);
+
+        const aggregatedDoc = DB.put.args[0][0];
+        chai.expect(aggregatedDoc._id).to.match(/telemetry-2018-9-greg/);
+        chai.expect(aggregatedDoc.metrics).to.deep.equal({
+          foo: 'stats',
+          bar: 'more stats',
+        });
+        chai.expect(aggregatedDoc.type).to.equal('telemetry');
+        chai.expect(aggregatedDoc.metadata.year).to.equal(2018);
+        chai.expect(aggregatedDoc.metadata.month).to.equal(9);
+        chai.expect(aggregatedDoc.metadata.user).to.equal('greg');
+        chai.expect(aggregatedDoc.dbInfo).to.deep.equal({ some: 'stats' });
+        chai.expect(aggregatedDoc.device).to.deep.equal({
+          userAgent: 'Agent Smith',
+          hardwareConcurrency: 4,
+          screen: {
+            width: 768,
+            height: 1024,
+          },
+        });
+      });
+    });
+
+    it('deals with 409s by making the ID unique and noting the conflict in the new document', () => {
```

Comment:
I guess we should all know by now that 409 is a conflict :slightly_frowning_face:

Reply:
No it's a good point, I'll remove the 409 reference and just talk about conflicts